File: Game-Dev-Math/QuadTree/quadtree.py
class Point:
    def __init__(self, x: float, y: float, user_data=None):
        self.x = x
        self.y = y
        self.user_data = user_data

class Rectangle:
    def __init__(self, x: float, y: float, w: float, h: float):
        # x, y is the center of the rectangle. w, h are the half-widths (distance from center to edge)
        self.x = x
        self.y = y
        self.w = w
        self.h = h

    def contains(self, point: Point) -> bool:
        return (self.x - self.w <= point.x <= self.x + self.w and
                self.y - self.h <= point.y <= self.y + self.h)

    def intersects(self, range_rect) -> bool:
        return not (range_rect.x - range_rect.w > self.x + self.w or
                    range_rect.x + range_rect.w < self.x - self.w or
                    range_rect.y - range_rect.h > self.y + self.h or
                    range_rect.y + range_rect.h < self.y - self.h)

class QuadTree:
    def __init__(self, boundary: Rectangle, capacity: int):
        self.boundary = boundary
        self.capacity = capacity
        self.points = []
        self.divided = False

    def subdivide(self):
        x, y, w, h = self.boundary.x, self.boundary.y, self.boundary.w, self.boundary.h
        
        ne = Rectangle(x + w/2, y - h/2, w/2, h/2)
        nw = Rectangle(x - w/2, y - h/2, w/2, h/2)
        se = Rectangle(x + w/2, y + h/2, w/2, h/2)
        sw = Rectangle(x - w/2, y + h/2, w/2, h/2)
        
        self.northeast = QuadTree(ne, self.capacity)
        self.northwest = QuadTree(nw, self.capacity)
        self.southeast = QuadTree(se, self.capacity)
        self.southwest = QuadTree(sw, self.capacity)
        
        self.divided = True

    def insert(self, point: Point) -> bool:
        if not self.boundary.contains(point):
            return False

        if len(self.points) < self.capacity:
            self.points.append(point)
            return True
        else:
            if not self.divided:
                self.subdivide()

            if self.northeast.insert(point): return True
            if self.northwest.insert(point): return True
            if self.southeast.insert(point): return True
            if self.southwest.insert(point): return True

    def query(self, range_rect: Rectangle, found: list = None) -> list:
        if found is None:
            found = []

        if not self.boundary.intersects(range_rect):
            return found

        for p in self.points:
            if range_rect.contains(p):
                found.append(p)

        if self.divided:
            self.northwest.query(range_rect, found)
            self.northeast.query(range_rect, found)
            self.southwest.query(range_rect, found)
            self.southeast.query(range_rect, found)

        return found

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # Create a 200x200 map centered at (100, 100)
    boundary = Rectangle(100, 100, 100, 100)
    qt = QuadTree(boundary, capacity=4)
    
    # Insert 5 points. The 5th point will force a subdivision!
    qt.insert(Point(50, 50))
    qt.insert(Point(60, 60))
    qt.insert(Point(40, 40))
    qt.insert(Point(45, 45))
    qt.insert(Point(150, 150)) # Triggers split
    
    # Query a small 40x40 box centered at (50, 50)
    query_box = Rectangle(50, 50, 20, 20)
    results = qt.query(query_box)
    
    assert len(results) == 4, f"Expected 4 points in query box, found {len(results)}"
    
    # Verify the distant point wasn't included
    for p in results:
        assert p.x != 150, "QuadTree math failed, included a point outside the query box!"
        
    print("Python QuadTree Spatial Partitioning Test Passed!")

File: Game-Dev-Math/QuadTree/quadtree.ts
class Point {
    constructor(public x: number, public y: number, public userData: any = null) {}
}

class Rectangle {
    // x, y is the center. w, h are half-widths.
    constructor(public x: number, public y: number, public w: number, public h: number) {}

    contains(point: Point): boolean {
        return (point.x >= this.x - this.w && point.x <= this.x + this.w &&
                point.y >= this.y - this.h && point.y <= this.y + this.h);
    }

    intersects(range: Rectangle): boolean {
        return !(range.x - range.w > this.x + this.w ||
                 range.x + range.w < this.x - this.w ||
                 range.y - range.h > this.y + this.h ||
                 range.y + range.h < this.y - this.h);
    }
}

export class QuadTree {
    private points: Point[] = [];
    private divided: boolean = false;
    private nw!: QuadTree;
    private ne!: QuadTree;
    private sw!: QuadTree;
    private se!: QuadTree;

    constructor(private boundary: Rectangle, private capacity: number) {}

    private subdivide(): void {
        const { x, y, w, h } = this.boundary;
        this.ne = new QuadTree(new Rectangle(x + w/2, y - h/2, w/2, h/2), this.capacity);
        this.nw = new QuadTree(new Rectangle(x - w/2, y - h/2, w/2, h/2), this.capacity);
        this.se = new QuadTree(new Rectangle(x + w/2, y + h/2, w/2, h/2), this.capacity);
        this.sw = new QuadTree(new Rectangle(x - w/2, y + h/2, w/2, h/2), this.capacity);
        this.divided = true;
    }

    insert(point: Point): boolean {
        if (!this.boundary.contains(point)) return false;

        if (this.points.length < this.capacity) {
            this.points.push(point);
            return true;
        }

        if (!this.divided) this.subdivide();

        if (this.ne.insert(point)) return true;
        if (this.nw.insert(point)) return true;
        if (this.se.insert(point)) return true;
        if (this.sw.insert(point)) return true;

        return false;
    }

    query(range: Rectangle, found: Point[] = []): Point[] {
        if (!this.boundary.intersects(range)) return found;

        for (const p of this.points) {
            if (range.contains(p)) found.push(p);
        }

        if (this.divided) {
            this.nw.query(range, found);
            this.ne.query(range, found);
            this.sw.query(range, found);
            this.se.query(range, found);
        }

        return found;
    }
}

// --- CI/CD Automated Test ---
const boundary = new Rectangle(100, 100, 100, 100);
const qt = new QuadTree(boundary, 4);

qt.insert(new Point(50, 50));
qt.insert(new Point(60, 60));
qt.insert(new Point(40, 40));
qt.insert(new Point(45, 45));
qt.insert(new Point(150, 150)); 

const queryBox = new Rectangle(50, 50, 20, 20);
const results = qt.query(queryBox);

const hasDistantPoint = results.some(p => p.x === 150);

if (results.length === 4 && !hasDistantPoint) {
    console.log("TypeScript QuadTree Spatial Partitioning Test Passed!");
} else {
    process.exit(1);
}